kbfstool/md_dump_common.go

```diff
@@ -0,0 +1,183 @@
+			// LoadUserPlusKeys call per user for a single
+			// replacements map.
+			if _, ok := replacements[u.String()]; ok {
+				continue
+			}
+
+			username, _, err := service.Resolve(
+				ctx, fmt.Sprintf("uid:%s", u))
+			if err == nil {
+				replacements[u.String()] = fmt.Sprintf(
+					"%s (uid:%s)", username, u)
+			} else {
+				replacements[u.String()] = fmt.Sprintf(
+					"<unknown username> (uid:%s)", u)
+				printError(prefix, err)
+			}
+
+			ui, err := service.LoadUserPlusKeys(ctx, u, "")
+			if err != nil {
+				printError(prefix, err)
+				continue
+			}
+
+			for _, k := range ui.CryptPublicKeys {
+				if _, ok := replacements[k.String()]; ok {
+					continue
+				}
+
+				if deviceStr, ok := mdDumpGetDeviceStringForCryptPublicKey(k, ui); ok {
+					replacements[k.String()] = deviceStr
+				}
+			}
+
+			for k := range ui.RevokedCryptPublicKeys {
+				if _, ok := replacements[k.String()]; ok {
+					continue
+				}
+
+				if deviceStr, ok := mdDumpGetDeviceStringForCryptPublicKey(k, ui); ok {
+					replacements[k.String()] = deviceStr
+				}
+			}
+
+			for _, k := range ui.VerifyingKeys {
+				if _, ok := replacements[k.String()]; ok {
+					continue
+				}
+
+				if deviceStr, ok := mdDumpGetDeviceStringForVerifyingKey(k, ui); ok {
+					replacements[k.String()] = deviceStr
+				}
+			}
+
+			for k := range ui.RevokedVerifyingKeys {
+				if _, ok := replacements[k.String()]; ok {
+					continue
+				}
+
+				if deviceStr, ok := mdDumpGetDeviceStringForVerifyingKey(k, ui); ok {
+					replacements[k.String()] = deviceStr
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func mdDumpReplaceAll(s string, replacements map[string]string) string {
+	for old, new := range replacements {
+		s = strings.Replace(s, old, new, -1)
+	}
+	return s
+}
+
+func mdDumpReadOnlyRMDWithReplacements(
+	ctx context.Context, codec kbfscodec.Codec,
+	replacements map[string]string, rmd libkbfs.ReadOnlyRootMetadata) error {
+	c := spew.NewDefaultConfig()
+	c.Indent = "  "
+	c.DisablePointerAddresses = true
+	c.DisableCapacities = true
+	c.SortKeys = true
+
+	fmt.Print("Root metadata\n")
+	fmt.Print("-------------\n")
+
+	brmdDump, err := kbfsmd.DumpRootMetadata(codec, rmd.GetBareRootMetadata())
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%s\n", mdDumpReplaceAll(brmdDump, replacements))
+
+	fmt.Print("Extra metadata\n")
+	fmt.Print("--------------\n")
+	extraDump, err := kbfsmd.DumpExtraMetadata(codec, rmd.Extra())
+	if err != nil {
+		return err
+	}
+	fmt.Printf("%s\n", mdDumpReplaceAll(extraDump, replacements))
+
+	fmt.Print("Private metadata\n")
+	fmt.Print("----------------\n")
+	pmdDump, err := libkbfs.DumpPrivateMetadata(
+		codec, len(rmd.GetSerializedPrivateMetadata()), *rmd.Data())
+	if err != nil {
+		return err
+	}
+	fmt.Printf("%s", mdDumpReplaceAll(pmdDump, replacements))
+
+	return nil
+}
+
+func mdDumpReadOnlyRMD(ctx context.Context, prefix string,
+	config libkbfs.Config, replacements map[string]string,
```

Comment:
Why does this function need to take `replacements`?  It seems like the callers of it don't need to use it after the dump, so might as well construct it within this function, right?

Reply:
It's so that calls for subsequent inputs reuse any entries filled in by this call; when I was running this, someone noticed elevated load on the API server, so decided to do caching here. Added `replacementMap` type with comment.